lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+      assert(subfieldLowering.hasValue() && "no subfield lowering for input");
+
+      // Map the result to the root bundle block argument, and append the
+      // suffix to what was already mapped.
+      auto subfieldInfo = subfieldLowering.getValue();
+      rootBundle = subfieldInfo.first;
+      suffix.assign(subfieldInfo.second);
+      suffix.push_back('_');
+      suffix.append(fieldname);
+    } else if (isa<InstanceOp>(owner)) {
+      // When the input is from an instance, look up the new instance.
+      Optional<Value> newInstance = getInstanceLowering(input);
+      assert(newInstance.hasValue() && "no instance lowering for input");
+
+      // Check if this subfield was originally a bundle type.
+      BundleType oldType = input.getType().cast<BundleType>();
+      FIRRTLType elementType = oldType.getElementType(fieldname);
+      assert(elementType && "no element found in lowered bundle");
+
+      if (elementType.isa<BundleType>()) {
+        // If it was a bundle, flatten it the usual way.
+        SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+        flattenBundleTypes(elementType, fieldname, false, fieldTypes);
+
+        for (auto field : fieldTypes) {
+          // Create a new subfield for the flattened field.
+          auto newSubfield =
+              builder->create<SubfieldOp>(op.getLoc(), getPortType(field),
+                                          newInstance.getValue(), field.suffix);
+
+          // Use the original result and the sub-field name (without the prefix)
+          // to remember the new subfield.
+          auto fieldName = StringRef(field.suffix).drop_front(fieldname.size());
+          setBundleLowering(result, fieldName, newSubfield);
+        }
+      }
+    }
+  }
+
+  Optional<Value> newValue = getBundleLowering(rootBundle, suffix);
+  if (newValue.hasValue()) {
+    // If the lowering for the current result exists, we have finished
+    // flattening. Replace the current result with the flattened value.
+    result.replaceAllUsesWith(newValue.getValue());
+  } else {
+    // Otherwise, remember the root bundle and the suffix so far.
+    setSubfieldLowering(result, rootBundle, std::string(suffix));
+  }
+
+  removeOp(op);
+
+  return success();
+}
+
+// Lowering connects only has to deal with one special case: connecting two
+// bundles. This situation should only arise when both of the arguments are a
+// bundle that was:
+//   a) originally a block argument
+//   b) originally an instance's port
+//
+// When two such bundles are connected, none of the subfield visits have a
+// chance to lower them, so we must ensure they have the same number of
+// flattened values and flatten out this connect into multiple connects.
+LogicalResult FIRRTLTypesLowering::visitStmt(ConnectOp op) {
+  Value lhs = op.lhs();
+  Value rhs = op.rhs();
+
+  // If we aren't connecting two bundles, there is nothing to do.
+  if (!(lhs.getType().isa<BundleType>() && rhs.getType().isa<BundleType>()))
```

Comment:
Please apply demorgans law to simplify the conditional.

What about outer level flip types?

Reply:
I'll simplify the conditional in the next revision.

I need to think about outer level flip types. This logic is intended to exit early unless we are in the very specific situation mentioned in the function comment. I don't think we would ever have outer level flip types in that situation, but agreed that I need to think through this.